### PR TITLE
Removing use of ManageWarnings within the code and unit tests.

### PR DESCRIPTION
**Pull request by Katie-Howard**
Closes https://github.com/metoppv/mo-blue-team/issues/437

This PR removes the `ManageWarnings` functionality usage from IMPROVER codebase as this handled centrally by Pytest. Before we adopted Pytest, this functionality was very useful with unittest.

Testing:
 - [x] Ran tests and they passed OK

### CODE REVIEW

File: improver_tests/blending/calculate_weights_and_blend/test_WeightAndBlend.py [hunk 1]
```diff
@@ -364,28 +360,13 @@ def setUp(self):
...
-        """Test a warning is raised if blending masked data with non-spatial
-        weights."""
-        ukv_cube = self.ukv_cube.copy(
-            data=np.ma.masked_where(self.ukv_cube.data < 0.5, self.ukv_cube.data)
-        )
-        self.plugin_cycle.process(
-            [ukv_cube, self.ukv_cube_latest], cycletime=self.cycletime,
-        )
-        message = "Blending masked data without spatial weights"
-        self.assertTrue(any(message in str(item) for item in warning_list))
```
Comment (bayliffe) [inline improver_tests/blending/calculate_weights_and_blend/test_WeightAndBlend.py:386]: Added back using pytest.warns.

File: improver_tests/blending/weighted_blend/test_MergeCubesForWeightedBlending.py [hunk 2]
```diff
@@ -74,17 +73,6 @@ def test_error_missing_model_id_attr(self):
...
-    @ManageWarnings(record=True)
-    def test_warning_unnecessary_model_id_attr(self, warning_list=None):
-        """Test warning if model_id_attr is set for non-model blending"""
-        warning_msg = "model_id_attr not required"
-        plugin = MergeCubesForWeightedBlending(
-            "realization", model_id_attr="mosg__model_configuration"
-        )
-        self.assertTrue(any(item.category == UserWarning for item in warning_list))
-        self.assertTrue(any(warning_msg in str(item) for item in warning_list))
-        self.assertIsNone(plugin.model_id_attr)
```
Comment (bayliffe) [inline improver_tests/blending/weighted_blend/test_MergeCubesForWeightedBlending.py:86]: Added back using pytest.warns.

File: improver_tests/calibration/ensemble_calibration/test_ContinuousRankedProbabilityScoreMinimisers.py [hunk 3]
```diff
@@ -559,74 +485,6 @@ def test_realizations_predictor_point_by_point(self):
...
-            self.forecast_predictor_mean,
-            self.truth,
-            self.forecast_variance,
-            distribution,
-        )
-        warning_msg_min = "Minimisation did not result in convergence after"
-        warning_msg_iter = "The final iteration resulted in a percentage "
-        self.assertTrue(any(item.category == UserWarning for item in warning_list))
-        self.assertTrue(any(warning_msg_min in str(item) for item in warning_list))
-        self.assertTrue(any(warning_msg_iter in str(item) for item in warning_list))
```
Comment (bayliffe) [inline improver_tests/calibration/ensemble_calibration/test_ContinuousRankedProbabilityScoreMinimisers.py:620]: Added back using pytest.warns

File: improver_tests/calibration/ensemble_calibration/test_ContinuousRankedProbabilityScoreMinimisers.py [hunk 4]
```diff
@@ -1023,67 +808,6 @@ def test_realizations_predictor_max_iterations(self):
...
```
Comment (bayliffe) [inline improver_tests/calibration/ensemble_calibration/test_ContinuousRankedProbabilityScoreMinimisers.py:1077]: Added back using pytest.warns

File: improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py [hunk 5]
```diff
@@ -247,43 +245,6 @@ def test_monotonic_case(self, warning_list=None):
...
-        self.forecast.coord(self.threshold).attributes[
-            "spp__relative_to_threshold"
-        ] = "below"
-
-        self.plugin._ensure_monotonicity_across_thresholds(self.forecast)
-
-        assert_array_equal(self.forecast.data, expected.data)
-        warning_msg = "Below threshold probabilities are not increasing"
-        self.assertTrue(any(item.category == UserWarning for item in warning_list))
-        self.assertTrue(any(warning_msg in str(item) for item in warning_list))
```
Comment (bayliffe) [inline improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py:285]: Added back using pytest.warns.

File: improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py [hunk 6]
```diff
@@ -473,51 +432,6 @@ def test_calibrating_forecast_with_reliability_table_cubelist(
...
-        assert_allclose(result[0].data, expected_0)
-        assert_allclose(result[1].data, expected_1)
-
-        warning_msg = (
-            "The following thresholds were not calibrated due to "
-            "insufficient forecast counts in reliability table "
-            "bins: [275.0]"
-        )
-        self.assertTrue(any(item.category == UserWarning for item in warning_list))
-        self.assertTrue(any(warning_msg in str(item) for item in warning_list))
```
Comment (bayliffe) [inline improver_tests/calibration/reliability_calibration/test_ApplyReliabilityCalibration.py:518]: Added back with pytest.warns

File: improver_tests/ensemble_copula_coupling/test_ConvertProbabilitiesToPercentiles.py [hunk 7]
```diff
@@ -125,31 +124,6 @@ def test_endpoints_of_distribution_exceeded(self):
...
-            "not in ascending order as required for the cumulative distribution "
-            "function (CDF). This is due to the threshold values exceeding "
-            "the range given by the ECC bounds (-40, 50). The threshold "
-            "points that have exceeded the existing bounds will be used as "
-            "new bounds."
-        )
-        plugin._add_bounds_to_thresholds_and_probabilities(
-            threshold_points, probabilities_for_cdf, self.bounds_pairing
-        )
-        self.assertTrue(any(warning_msg in str(item) for item in warning_list))
```
Comment (bayliffe) [inline improver_tests/ensemble_copula_coupling/test_ConvertProbabilitiesToPercentiles.py:150]: Added back using pytest.warns

File: improver_tests/ensemble_copula_coupling/test_ConvertProbabilitiesToPercentiles.py [hunk 8]
```diff
@@ -325,25 +299,6 @@ def test_check_data_multiple_timesteps(self):
...
-        data = data[:, np.newaxis, np.newaxis]
-        cube = set_up_probability_cube(
-            data.astype(np.float32), ECC_TEMPERATURE_THRESHOLDS, threshold_units="degC"
-        )
-
-        warning_msg = "The probability values used to construct the"
-        Plugin()._probabilities_to_percentiles(
-            cube, self.percentiles, self.bounds_pairing
-        )
-        self.assertTrue(any(warning_msg in str(item) for item in warning_list))
```
Comment (bayliffe) [inline improver_tests/ensemble_copula_coupling/test_ConvertProbabilitiesToPercentiles.py:345]: Added back using pytest.warns

File: improver_tests/generate_ancillaries/test_GenerateTopographicZoneWeights.py [hunk 9]
```diff
@@ -421,30 +420,6 @@ def test_one_band_with_orography_in_band(self):
...
-    def test_warning_if_orography_below_bands(self, warning_list=None):
-        """Test that a warning is raised if the orography is lower than the
-        minimum band."""
-        orography_data = np.array([[60.0, 70.0], [80.0, 90.0]])
-        orography = self.orography.copy(data=orography_data)
-        thresholds_dict = {"bounds": [[100, 150]], "units": "m"}
-        msg = "The minimum orography is lower than the lowest band"
-        self.plugin.process(orography, thresholds_dict, self.landmask)
-        self.assertTrue(any(item.category == UserWarning for item in warning_list))
-        self.assertTrue(any(msg in str(item) for item in warning_list))
```
Comment (bayliffe) [inline improver_tests/generate_ancillaries/test_GenerateTopographicZoneWeights.py:446]: Added back using pytest.warns.

File: improver_tests/metadata/test_forecast_times.py [hunk 10]
```diff
@@ -179,19 +178,6 @@ def test_check_time_unit_conversion(self):
...
-        """Test a warning is raised if the calculated forecast period is
-        negative"""
-        # default cube has a 4 hour forecast period, so add 5 hours to frt
-        self.frt_coord.points = self.frt_coord.points + 5 * 3600
-        result = _calculate_forecast_period(self.time_coord, self.frt_coord)
-        warning_msg = "The values for the time"
-        result = _calculate_forecast_period(self.time_coord, self.frt_coord)
-        self.assertTrue(any(item.category == UserWarning for item in warning_list))
-        self.assertTrue(any(warning_msg in str(item) for item in warning_list))
-        self.assertEqual(result.points, [-3600])
```
Comment (bayliffe) [inline improver_tests/metadata/test_forecast_times.py:193]: Added back using pytests.warn
Removed duplicate calculation of result.

File: improver_tests/nbhood/recursive_filter/test_RecursiveFilter.py [hunk 11]
```diff
@@ -169,19 +168,6 @@ def test_iterations(self):
...
-        """Test when the iteration value is more than 3 it warns."""
-        iterations = 5
-        warning_msg = (
-            "More than two iterations degrades the conservation"
-            "of probability assumption."
-        )
-
-        RecursiveFilter(iterations=iterations)
-        self.assertTrue(any(item.category == UserWarning for item in warning_list))
-        self.assertTrue(any(warning_msg in str(item) for item in warning_list))
```
Comment (bayliffe) [inline improver_tests/nbhood/recursive_filter/test_RecursiveFilter.py:183]: Added back using pytest.warns

File: improver_tests/nowcasting/optical_flow/test_OpticalFlow.py [hunk 12]
```diff
@@ -541,80 +540,6 @@ def test_values(self):
...
-        )
-        self.assertTrue(len(warning_list) == 0)
-
-    @ManageWarnings(record=True)
-    def test_no_warning_raised_outside_rain(self, warning_list=None):
-        """Test warning ignores zeros outside the rain area mask"""
-        rain = np.array([[0, 0, 1], [0, 1, 1], [1, 1, 1]])
-        wind = np.array([[0, 0, 1], [0, 1, 1], [1, 1, 1]])
-        self.plugin._zero_advection_velocities_warning(wind, np.where(rain > 0))
-        self.assertTrue(len(warning_list) == 0)
```
Comment (bayliffe) [inline improver_tests/nowcasting/optical_flow/test_OpticalFlow.py:615]: Added back in using following syntax:
```
        with warnings.catch_warnings():
            warnings.simplefilter(action="error", category=UserWarning)
```
Such that an error is raised if a UserWarning is raised and the test fails.

File: improver_tests/spotdata/test_NeighbourSelection.py [hunk 13]
```diff
@@ -602,26 +544,6 @@ def test_all_invalid_points(self):
...
-        distance = np.arange(5)
-        indices = np.arange(5)
-
-        plugin.select_minimum_dz(
-            self.region_orography, site_altitude, nodes, distance, indices
-        )
-
-        msg = "Limit on number of nearest neighbours"
-        self.assertTrue(any(msg in str(warning) for warning in warning_list))
-        self.assertTrue(any(item.category == UserWarning for item in warning_list))
```
Comment (bayliffe) [inline improver_tests/spotdata/test_NeighbourSelection.py:623]: Added back using pytest.warns

File: improver_tests/utilities/cube_manipulation/test_compare_attributes.py [hunk 14]
```diff
@@ -69,15 +68,6 @@ def test_basic(self):
...
         self.assertArrayEqual(result, [{}, {}])
 
-    @ManageWarnings(record=True)
-    def test_warning(self, warning_list=None):
-        """Test that the utility returns warning if only one cube supplied."""
-        result = compare_attributes(iris.cube.CubeList([self.cube]))
-        self.assertTrue(any(item.category == UserWarning for item in warning_list))
-        warning_msg = "Only a single cube so no differences will be found "
-        self.assertTrue(any(warning_msg in str(item) for item in warning_list))
-        self.assertArrayEqual(result, [])
```
Comment (bayliffe) [inline improver_tests/utilities/cube_manipulation/test_compare_attributes.py:79]: Added back using pytest.warns

File: improver_tests/utilities/cube_manipulation/test_compare_coords.py [hunk 15]
```diff
@@ -66,16 +65,6 @@ def test_basic(self):
...
 
-    @ManageWarnings(record=True)
-    def test_catch_warning(self, warning_list=None):
-        """Test warning is raised if the input is cubelist of length 1."""
-        cube = self.cube.copy()
-        result = compare_coords(iris.cube.CubeList([cube]))
-        self.assertTrue(any(item.category == UserWarning for item in warning_list))
-        warning_msg = "Only a single cube so no differences will be found "
-        self.assertTrue(any(warning_msg in str(item) for item in warning_list))
-        self.assertEqual(result, [])
```
Comment (bayliffe) [inline improver_tests/utilities/cube_manipulation/test_compare_coords.py:77]: Added back using pytest.warns

File: improver_tests/utilities/cube_manipulation/test_sort_coord_in_cube.py [hunk 16]
```diff
@@ -169,19 +168,6 @@ def test_latitude(self):
...
-        """Test that a warning is successfully raised when circular
-        coordinates are sorted."""
-        self.ascending_cube.data[:, 0, 0] = 6.0
-        coord_name = "latitude"
-        self.ascending_cube.coord(coord_name).circular = True
-        result = sort_coord_in_cube(self.ascending_cube, coord_name, descending=True)
-        self.assertTrue(any(item.category == UserWarning for item in warning_list))
-        warning_msg = "The latitude coordinate is circular."
-        self.assertTrue(any(warning_msg in str(item) for item in warning_list))
-        self.assertIsInstance(result, iris.cube.Cube)
```
Comment (bayliffe) [inline improver_tests/utilities/cube_manipulation/test_sort_coord_in_cube.py:183]: Added back using pytest.warns

File: improver_tests/utilities/temporal/test_temporal.py [hunk 17]
```diff
@@ -315,27 +314,13 @@ def test_valid_time(self):
...
-        plugin = extract_cube_at_time
-        time_dt = datetime(2017, 2, 18, 6, 0)
-        time_constraint = iris.Constraint(
-            time=PartialDateTime(time_dt.year, time_dt.month, time_dt.day, time_dt.hour)
-        )
-        cubes = CubeList([self.cube])
-        plugin(cubes, time_dt, time_constraint)
-        warning_msg = "Forecast time"
-        self.assertTrue(any(item.category == UserWarning for item in warning_list))
-        self.assertTrue(any(warning_msg in str(item) for item in warning_list))
```
Comment (bayliffe) [inline improver_tests/utilities/temporal/test_temporal.py:337]: Added back using pytest.warns

File: improver_tests/utilities/test_InterpolateUsingDifference.py [hunk 18]
```diff
@@ -277,21 +276,6 @@ def test_linear_failure(self):
...
-
-        self.sleet_rain.data = np.ones((3, 3), dtype=np.float32)
-        expected = self.sleet_rain.copy()
-        warning_msg = "Input cube unmasked, no data to fill in, returning"
-
-        result = InterpolateUsingDifference().process(self.sleet_rain, self.snow_sleet)
-
-        self.assertEqual(result, expected)
-        self.assertTrue(any(item.category == UserWarning for item in warning_list))
-        self.assertTrue(any(warning_msg in str(item) for item in warning_list))
```
Comment (bayliffe) [inline improver_tests/utilities/test_InterpolateUsingDifference.py:293]: Added back using pytest.warns

File: improver_tests/wind_calculations/wind_gust_diagnostic/test_WindGustDiagnostic.py [hunk 19]
```diff
@@ -162,19 +161,6 @@ def test_fails_if_data_is_not_cube(self):
...
-        """Test it raises a warning if standard names do not match."""
-        plugin = WindGustDiagnostic(self.wg_perc, self.ws_perc)
-        warning_msg = "Warning mismatching name for data expecting"
-        result, perc_coord = plugin.extract_percentile_data(
-            self.cube_wg, self.wg_perc, "wind_speed"
-        )
-        self.assertTrue(any(item.category == UserWarning for item in warning_list))
-        self.assertTrue(any(warning_msg in str(item) for item in warning_list))
-        self.assertIsInstance(result, Cube)
-        self.assertIsInstance(perc_coord, iris.coords.Coord)
```
Comment (bayliffe) [inline improver_tests/wind_calculations/wind_gust_diagnostic/test_WindGustDiagnostic.py:176]: Added back using pytest.warns

**Review by bayliffe — DISMISSED** [review]
I've reviewed this in as much as I have checked the removals are as expected. I have reintroduced tests that were removed erroneously, and I have fixed formatting errors. This needs a thorough second review in light of the changes I have made.

File: improver_tests/nowcasting/optical_flow/test_OpticalFlow.py [hunk 20]
```diff
@@ -911,16 +913,14 @@ def test_error_nonmatching_inputs(self):
...
         """Test code raises a warning if input cubes have
         non-rain variable names"""
         self.cube1.rename("snowfall_rate")
         self.cube2.rename("snowfall_rate")
-        _, _ = self.plugin.process(self.cube1, self.cube2, boxsize=3)
         warning_msg = "Input data are of non-precipitation type"
-        self.assertTrue(any(item.category == UserWarning for item in warning_list))
-        self.assertTrue(any(warning_msg in str(item) for item in warning_list))
+        with pytest.warns(UserWarning, match=warning_msg):
+            _, _ = self.plugin.process(self.cube1, self.cube2, boxsize=3)
```
Comment (MoseleyS) [inline improver_tests/nowcasting/optical_flow/test_OpticalFlow.py:923]: You've done this exactly as it was before, but it should be simplified to:
```suggestion
            self.plugin.process(self.cube1, self.cube2, boxsize=3)
```

File: improver_tests/regrid/test_AdjustLandSeaPoints.py [hunk 21]
```diff
@@ -45,7 +45,6 @@
     set_up_variable_cube,
 )
 from improver.utilities.spatial import OccurrenceWithinVicinity
-from improver.utilities.warnings_handler import ManageWarnings
 
 # The warning messages are internal to the iris.analysis module v2.2.0
 IGNORED_MESSAGES = ["Using a non-tuple sequence for multidimensional indexing"]
```
Comment (MoseleyS) [inline improver_tests/regrid/test_AdjustLandSeaPoints.py:50]: `IGNORED_MESSAGES` and `WARNING_TYPES` are no longer used and need to be deleted.

**Review by MoseleyS — CHANGES_REQUESTED** [review]
Looks very good. Spotted a couple of things to fix.

**Review by MoseleyS — APPROVED** [review]

**Pull request merged**